blog/2021-11-04-introducing-pwa.md

```diff
@@ -0,0 +1,168 @@
+
+### “There’s an update available for this app”
+
+When using PWA, the app will need to handle updates to the app in a special way. Because the scripts and assets that make up the app are precached and being served straight from the offline caches without checking the files on the server, in order for a new version of an app that’s deployed on the DHIS2 server to become available in a user’s browser, the _service worker_ in the user’s browser must update so that it can download the _new_ app’s files and serve those from the cache.
+
+To handle these situations, we have set up a system in the App Platform that notifies users of these updates and prompts them to reload the app to use them. The first time a service worker is installed in a PWA-enabled app, an Alert will be shown saying “There’s an update available for this app”, with the options “Update now” and “Not now”. Clicking “Update now” will reload all the open tabs of that app, and there will be a confirmation dialog before reloading if there are multiple tabs open to avoid losing unsaved data. Once the app has reloaded, offline capability will be available.
+
+![PWA Update prompt](./assets/introducing-pwa/pwa-update-prompt.png)
+
+A similar process will happen if a new version of the app is available on the server: the same prompt will show with the same options. When the app reloads after clicking “Update now”, the new version of the app will be in use. If “Not now” is clicked, the prompt will be shown again when the page reloads. An app update will be activated automatically if all the tabs of that app have been closed and then a new one is opened.
+
+If possible, it’s best to encourage users to save their data and update the app as soon as possible to make sure they’re using the latest version.
+
+### Cacheable Sections
+
+This is a set of features designed with the Dashboards app in mind to enable saving individual dashboards in the offline cache while leaving other dashboards uncached. They are designed in a generalized way to save any chunks of content with a React API, so you can use them if they suit your app too.
+
+![Dashboards app](./assets/introducing-pwa/dashboard-save-offline.png)
+
+The Dashboards app available in DHIS2 version 2.37 (viewable at https://play.dhis2.org/2.37dev) can be used an example of how cacheable sections work: when you visit the app and confirm any update prompts, you can go offline and still view the app. If you try to visit another dashboard though, you’ll see a “not available offline” notice.
+
+![Dashboard not available offline](./assets/introducing-pwa/dashboard-not-available-offline.png)
+
+The dashboard here is the **cacheable section** – it’s only made available offline when you do so explicitly. The rest of the app that continues to work while offline is called the **app shell**, and includes content like the scripts that run the app’s behavior, some user and server data, and the user’s list of dashboards.
```

Comment:
minor: we already use the term **app shell** to refer to the common wrapper provided by the platform.  Using this terminology again to refer to something a bit different might be confusing...

Reply:
I was thinking that we can avoid giving it a name and gloss over it a bit:

```suggestion
The dashboard here is the **cacheable section** – it’s only made available offline when you do so explicitly. The rest of the app that continues to work while offline includes content like the scripts that run the app’s behavior, some user and server data, and the user’s list of dashboards.
```

Reply:
Yeah it is regrettable to have that terminology conflict; I've been using "app shell" in the caching context because it seems like that's how people refer to "the static assets and data needed to run the core of the app" in PWA contexts. If you think we need a new name for it, we could work on a better one, but will have to rename a few [config options](https://github.com/dhis2/app-platform/blob/1d0423e135b71d2005198287075e47d939040049/cli/config/d2.pwa.config.js#L25) and change some documentation

What do you think about adding some more description to make it clearer that "app shell" here is referring to the cache contents, not to be confused with the common wrapper? Or removing the reference to "app shell" here like Viktor suggests?